copy/copy.go

```diff
@@ -489,8 +497,20 @@ func (ic *imageCopier) copyBlobFromStream(srcStream io.Reader, srcInfo types.Blo
 		inputInfo.Size = -1
 	}
 
+	var stream io.Reader
+
 	// === Finally, send the layer stream to dest.
-	uploadedInfo, err := ic.dest.PutBlob(destStream, inputInfo)
+	// the pipe is here to facilitate the stream copier which will act as an
+	// intermediary to signal progress information to any registered channel.
+	if ic.signal != nil && ic.signalInterval > 0 {
+		r, w := io.Pipe()
+		go streamcopy.WithProgress(w, destStream, inputInfo, ic.signalInterval, ic.signal)
+		stream = r
+	} else {
+		stream = destStream
+	}
```

Comment:
Please use the same stage description structure as the rest of the steps, this function is pretty complex and the regularity makes it easier to understand:

- Create a separate newline-separated section for this stage in the pipeline, starting with a `// === `-prefixed comment
- Use `destStream` as both the input and output of that pipeline stage (i.e. `if ic.signal != nil … { … destStream = r }  ` with no `else` branch)